File: server/app/routers/widget.py
from fastapi import APIRouter, Query, Request, Response, status

from app.db import SessionDep
from app.deps import Visitor, client_ip
from app.models import INBOUND, Conversation, Message
from app.schemas.widget import (
    ChatMessage,
    SendRequest,
    SessionOut,
    SessionRequest,
    ThreadDetail,
    ThreadList,
    ThreadOut,
    VisitorOut,
)
from app.schemas.ws import TicketOut
from app.services import events, ratelimit, tickets
from app.services import visitor as visitor_tokens
from app.services import widget as service
from app.services.security import utcnow

router = APIRouter(prefix="/api/widget", tags=["widget"])


def thread_out(conversation: Conversation, opening: str, unread: int) -> ThreadOut:
    return ThreadOut(
        id=conversation.id,
        status=conversation.status,
        title=conversation.subject or opening[: service.TITLE_LENGTH] or "New conversation",
        preview=conversation.last_message_preview or opening,
        unread=unread,
        last_at=conversation.last_message_at,
    )


def message_out(message: Message, authors: dict[int, str]) -> ChatMessage:
    return ChatMessage(
        id=message.id,
        seq=message.seq,
        sender="customer" if message.direction == INBOUND else "agent",
        author=authors.get(message.author_user_id) if message.author_user_id else None,
        body=message.body_text,
        at=message.created_at,
        client_msg_id=message.client_msg_id,
    )


@router.post("/session", response_model=SessionOut)
async def open_session(request: Request, body: SessionRequest, db: SessionDep) -> SessionOut:
    key = body.key.strip()
    ratelimit.enforce(ratelimit.WIDGET_SESSION, f"{client_ip(request)}:{key}")

    now = utcnow()
    workspace = await service.workspace_for_key(db, key)
    known = visitor_tokens.read_browser_id(body.browser_id, now=now) if body.browser_id else None

    identified = await service.identify(
        db,
        workspace,
        known_visitor_id=known,
        name=body.name,
        email=str(body.email) if body.email else None,
        now=now,
    )
    await db.commit()

    customer = identified.customer
    return SessionOut(
        session=visitor_tokens.sign_session(
            visitor_id=identified.visitor_id,
            customer_id=customer.id,
            workspace_id=workspace.id,
            now=now,
        ),
        browser_id=visitor_tokens.sign_browser_id(identified.visitor_id, now=now),
        workspace_name=workspace.name,
        greeting=workspace.widget_greeting,
        visitor=VisitorOut(id=customer.id, name=customer.name, email=customer.email),
    )


@router.post("/ws/ticket", response_model=TicketOut)
async def chat_ticket(visitor: Visitor) -> TicketOut:
    ratelimit.enforce(ratelimit.WS_TICKET, str(visitor.customer.id))
    claim = tickets.Claim(
        kind=tickets.VISITOR,
        workspace_id=visitor.workspace_id,
        customer_id=visitor.customer.id,
    )
    return TicketOut(ticket=tickets.mint(claim), expires_in=tickets.TTL_SECONDS)


@router.get("/conversations", response_model=ThreadList)
async def list_threads(visitor: Visitor) -> ThreadList:
    rows = await service.conversations_for(visitor.db, visitor.customer)
    return ThreadList(items=[thread_out(row.conversation, row.opening, row.unread) for row in rows])


@router.post("/conversations", response_model=ThreadDetail, status_code=status.HTTP_201_CREATED)
async def start_thread(visitor: Visitor, body: SendRequest) -> ThreadDetail:
    ratelimit.enforce(ratelimit.WIDGET_START, str(visitor.customer.id))

    now = utcnow()
    conversation, message = await service.start_with_message(
        visitor.db,
        visitor.customer,
        body=body.body,
        client_msg_id=body.client_msg_id,
        now=now,
    )
    await visitor.db.commit()
    events.message_saved(conversation, message)

    return ThreadDetail(
        thread=thread_out(conversation, message.body_text, 0),
        messages=[message_out(message, {})],
    )


@router.get("/conversations/{conversation_id}", response_model=ThreadDetail)
async def read_thread(
    visitor: Visitor,
    conversation_id: int,
    after_seq: int = Query(default=0, ge=0),
) -> ThreadDetail:
    conversation = await service.conversation_of(visitor.db, visitor.customer, conversation_id)
    messages = await service.messages_of(visitor.db, conversation, after_seq=after_seq)
    authors = await service.authors_of(visitor.db, messages)
    unread = await service.unread_for(visitor.db, conversation)
    opening = await service.opening_of(visitor.db, conversation)

    return ThreadDetail(
        thread=thread_out(conversation, opening, unread),
        messages=[message_out(message, authors) for message in messages],
    )


@router.post("/conversations/{conversation_id}/messages", response_model=ChatMessage)
async def send_message(visitor: Visitor, conversation_id: int, body: SendRequest) -> ChatMessage:
    ratelimit.enforce(ratelimit.WIDGET_SEND, str(visitor.customer.id))

    conversation = await service.conversation_of(visitor.db, visitor.customer, conversation_id)
    message = await service.send(
        visitor.db,
        conversation,
        body=body.body,
        client_msg_id=body.client_msg_id,
        now=utcnow(),
    )
    await visitor.db.commit()
    events.message_saved(conversation, message)
    return message_out(message, {})


@router.post("/conversations/{conversation_id}/read", status_code=status.HTTP_204_NO_CONTENT)
async def mark_read(visitor: Visitor, conversation_id: int) -> Response:
    conversation = await service.conversation_of(visitor.db, visitor.customer, conversation_id)
    await service.mark_agent_replies_read(visitor.db, conversation, utcnow())
    await visitor.db.commit()
    events.read_by(conversation, who=events.CUSTOMER)
    return Response(status_code=status.HTTP_204_NO_CONTENT)

File: client/lib/widget.ts
import { toError } from "@/lib/api";
import { type Live, type LiveEvent, keepConnected } from "@/lib/live";

export type ChatMessage = {
  id: number;
  seq: number;
  sender: "customer" | "agent";
  author: string | null;
  body: string;
  at: string;
  client_msg_id: string | null;
};

export type Thread = {
  id: number;
  status: string;
  title: string;
  preview: string;
  unread: number;
  last_at: string;
};

export type ThreadDetail = { thread: Thread; messages: ChatMessage[] };

export type Session = {
  session: string;
  browser_id: string;
  workspace_name: string;
  greeting: string | null;
  visitor: { id: number; name: string | null; email: string | null };
};

export const SESSION_GONE = "chat_session_gone";

const BROWSER_ID = "intercom.browser";

function remembered(key: string): string | null {
  try {
    return window.localStorage.getItem(`${BROWSER_ID}.${key}`);
  } catch {
    return null;
  }
}

function remember(key: string, browserId: string): void {
  try {
    window.localStorage.setItem(`${BROWSER_ID}.${key}`, browserId);
  } catch {
    return;
  }
}

async function call<T>(
  token: string | null,
  method: string,
  path: string,
  body?: unknown,
): Promise<T> {
  const headers: Record<string, string> = {};
  if (token) headers.authorization = `Bearer ${token}`;
  if (body !== undefined) headers["content-type"] = "application/json";

  const response = await fetch(`/api/widget${path}`, {
    method,
    credentials: "omit",
    cache: "no-store",
    headers,
    body: body === undefined ? undefined : JSON.stringify(body),
  });

  if (!response.ok) throw await toError(response);
  if (response.status === 204) return undefined as T;
  return (await response.json()) as T;
}

export async function openSession(
  key: string,
  who?: { name: string; email: string },
): Promise<Session> {
  const session = await call<Session>(null, "POST", "/session", {
    key,
    browser_id: remembered(key),
    name: who?.name,
    email: who?.email,
  });
  remember(key, session.browser_id);
  return session;
}

export function listThreads(token: string): Promise<{ items: Thread[] }> {
  return call<{ items: Thread[] }>(token, "GET", "/conversations");
}

export function readThread(token: string, id: number, afterSeq = 0): Promise<ThreadDetail> {
  const from = afterSeq > 0 ? `?after_seq=${afterSeq}` : "";
  return call<ThreadDetail>(token, "GET", `/conversations/${id}${from}`);
}

export function startThread(
  token: string,
  body: string,
  clientMsgId: string,
): Promise<ThreadDetail> {
  return call<ThreadDetail>(token, "POST", "/conversations", {
    body,
    client_msg_id: clientMsgId,
  });
}

export function sendMessage(
  token: string,
  id: number,
  body: string,
  clientMsgId: string,
): Promise<ChatMessage> {
  return call<ChatMessage>(token, "POST", `/conversations/${id}/messages`, {
    body,
    client_msg_id: clientMsgId,
  });
}

export function markRead(token: string, id: number): Promise<void> {
  return call<void>(token, "POST", `/conversations/${id}/read`);
}

export function newMessageId(): string {
  return crypto.randomUUID();
}

export function mergeBySeq(held: ChatMessage[], arriving: ChatMessage[]): ChatMessage[] {
  if (arriving.length === 0) return held;
  const bySeq = new Map(held.map((m) => [m.seq, m]));
  for (const message of arriving) bySeq.set(message.seq, message);
  return [...bySeq.values()].sort((a, b) => a.seq - b.seq);
}

export function highestSeq(messages: ChatMessage[]): number {
  return messages.reduce((top, m) => (m.seq > top ? m.seq : top), 0);
}

export function stillPending(pending: ChatMessage[], canonical: ChatMessage[]): ChatMessage[] {
  if (pending.length === 0) return pending;
  const landed = new Set(canonical.map((m) => m.client_msg_id).filter(Boolean));
  return pending.filter((m) => !landed.has(m.client_msg_id));
}

export function panelTicket(token: string): Promise<string> {
  return call<{ ticket: string }>(token, "POST", "/ws/ticket").then((r) => r.ticket);
}

export function connectPanel(
  getTicket: () => Promise<string>,
  onEvent: (event: LiveEvent) => void,
): Live {
  return keepConnected("/ws/widget", getTicket, onEvent);
}
